File: app/src/main/java/com/majeur/psclient/SignInDialog.java
package com.majeur.psclient;

import android.content.Context;
import android.os.Bundle;
import android.text.Editable;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.inputmethod.EditorInfo;
import android.widget.Button;
import android.widget.EditText;
import android.widget.TextView;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.DialogFragment;
import com.google.android.material.textfield.TextInputLayout;
import com.majeur.psclient.service.ShowdownService;
import com.majeur.psclient.util.SimpleTextWatcher;

public class SignInDialog extends DialogFragment implements View.OnClickListener, ShowdownService.AttemptSignInCallback {

    public static final String FRAGMENT_TAG = "sign-in-dialog";

    public static SignInDialog newInstance() {
        return new SignInDialog();
    }

    private ShowdownService mService;

    private EditText mUsernameEditText;
    private TextInputLayout mUsernameInputLayout;
    private EditText mPasswordEditText;
    private TextInputLayout mPasswordInputLayout;
    private Button mSignInButton;

    private boolean mHasView;
    private boolean mIsRequiringPassword;

    @Override
    public void onAttach(@NonNull Context context) {
        super.onAttach(context);
        mService = ((MainActivity) context).getService();
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        mHasView = true;
        return inflater.inflate(R.layout.dialog_sign_in, container, false);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        mUsernameEditText = view.findViewById(R.id.edit_text_username);
        //mUsernameEditText.setFilters(new InputFilter[] {new UserNameFilter(), new InputFilter.LengthFilter(18)});
        mUsernameEditText.addTextChangedListener(new SimpleTextWatcher() {
            @Override
            public void afterTextChanged(Editable editable) {
                if (editable.toString().matches(".*[\\|,;].*"))
                    mUsernameInputLayout.setError("| , ; are not valid characters in names");
                else
                    mUsernameInputLayout.setErrorEnabled(false);

            }
        });
        mUsernameEditText.requestFocus();
        mUsernameEditText.setOnEditorActionListener(mEnterPressListener);
        mUsernameInputLayout = (TextInputLayout) mUsernameEditText.getParent().getParent(); // TextInputLayout wraps its child in a FrameLayout

        mPasswordEditText = view.findViewById(R.id.edit_text_password);
        mPasswordEditText.setOnEditorActionListener(mEnterPressListener);
        mPasswordInputLayout = (TextInputLayout) mPasswordEditText.getParent().getParent();

        mSignInButton = view.findViewById(R.id.button_signin);
        mSignInButton.setOnClickListener(this);
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        mHasView = false;
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        mService = null;
    }

    @Override
    public void onClick(View view) {
        if (mIsRequiringPassword) {
            if (mPasswordEditText.getText().length() > 0) {
                mPasswordEditText.setEnabled(false);
                mSignInButton.setText("Loading...");
                mSignInButton.setEnabled(false);
                setCancelable(false);
                mService.attemptSignIn(mUsernameEditText.getText().toString(),
                                mPasswordEditText.getText().toString(), SignInDialog.this);

            }
        } else {
            if (mUsernameEditText.getText().length() > 0) {
                mUsernameInputLayout.setEnabled(false);
                mSignInButton.setText("Loading...");
                mSignInButton.setEnabled(false);
                setCancelable(false);
                mService.attemptSignIn(mUsernameEditText.getText().toString(), SignInDialog.this);
            }
        }
    }

    @Override
    public void onSuccess() {
        if (!mHasView) return;
        setCancelable(true);
        mUsernameInputLayout.setErrorEnabled(false);
        mPasswordInputLayout.setErrorEnabled(false);
        requireDialog().dismiss();
    }

    @Override
    public void onAuthenticationRequired() {
        if (!mHasView) return;
        setCancelable(true);
        mPasswordInputLayout.setVisibility(View.VISIBLE);
        mPasswordEditText.setEnabled(true);
        mSignInButton.setText("Sign in");
        mSignInButton.setEnabled(true);
        mIsRequiringPassword = true;
        mUsernameInputLayout.setErrorEnabled(false); // If any issue happened before
    }

    @Override
    public void onError(String reason) {
        if (!mHasView) return;
        setCancelable(true);
        if (mIsRequiringPassword) {
            mPasswordEditText.setEnabled(true);
            mSignInButton.setText("Sign in");
            mSignInButton.setEnabled(true);
            mPasswordInputLayout.setError("Wrong password, please try again.");
        } else {
            mUsernameInputLayout.setError(reason);
            mUsernameInputLayout.setEnabled(true);
            mSignInButton.setText("Sign in");
            mSignInButton.setEnabled(true);
        }
    }

    private final TextView.OnEditorActionListener mEnterPressListener = (v, actionId, event) -> {
        if (actionId == EditorInfo.IME_ACTION_GO) {
            onClick(v);
            return true;
        }
        return false;
    };
}

File: psclient/src/main/java/com/majeur/psclient/ui/SignInDialog.kt
package com.majeur.psclient.ui

import android.content.Context
import android.os.Bundle
import android.text.Editable
import android.view.KeyEvent
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.inputmethod.EditorInfo
import android.widget.TextView
import android.widget.TextView.OnEditorActionListener
import androidx.fragment.app.DialogFragment
import com.majeur.psclient.databinding.DialogSignInBinding
import com.majeur.psclient.service.ShowdownService
import com.majeur.psclient.service.ShowdownService.AttemptSignInCallback
import com.majeur.psclient.util.SimpleTextWatcher
import kotlinx.android.synthetic.main.dialog_sign_in.*

class SignInDialog : DialogFragment(), View.OnClickListener, AttemptSignInCallback {

    private var service: ShowdownService? = null
    private var requirePassword = false

    private var _binding: DialogSignInBinding? = null
    private val binding get() = _binding!!

    override fun onAttach(context: Context) {
        super.onAttach(context)
        service = (context as MainActivity).service
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        _binding = DialogSignInBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        binding.apply {
            username.addTextChangedListener(object : SimpleTextWatcher() {
                override fun afterTextChanged(editable: Editable) {
                    if (editable.matches(NAME_REGEX))
                        usernameContainer.error = "| , ; are not valid characters in names"
                    else usernameContainer.isErrorEnabled = false
                }
            })
            username.requestFocus()
            // username.setFilters(new InputFilter[] {new UserNameFilter(), new InputFilter.LengthFilter(18)});
            username.setOnEditorActionListener(mEnterPressListener)
            password.setOnEditorActionListener(mEnterPressListener)
            button.setOnClickListener(this@SignInDialog)
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        service = null
    }

    override fun onClick(view: View) {
        if (requirePassword) {
            if (binding.password.text.isNotEmpty()) {
                binding.password.isEnabled = false
                button.text = "Loading..."
                button.isEnabled = false
                isCancelable = false
                service?.attemptSignIn(binding.username.text.toString(),
                        binding.password.text.toString(), this@SignInDialog)
            }
        } else {
            if (binding.username.text.isNotEmpty()) {
                binding.username.isEnabled = false
                binding.button.text = "Loading..."
                button.isEnabled = false
                isCancelable = false
                service!!.attemptSignIn(binding.username.text.toString(), this@SignInDialog)
            }
        }
    }

    override fun onSuccess() {
        if (_binding == null) return
        isCancelable = true
        binding.usernameContainer.isErrorEnabled = false
        binding.passwordContainer.isErrorEnabled = false
        requireDialog().dismiss()
    }

    override fun onAuthenticationRequired() {
        if (_binding == null) return
        isCancelable = true
        requirePassword = true
        binding.apply {
            passwordContainer.visibility = View.VISIBLE
            password.isEnabled = true
            button.text = "Sign in"
            button.isEnabled = true
            usernameContainer.isErrorEnabled = false // If any issue happened before
        }
    }

    override fun onError(reason: String) {
        if (_binding == null) return
        isCancelable = true
        if (requirePassword) {
            binding.password.isEnabled = true
            binding.passwordContainer.error = reason
            binding.button.text = "Sign in"
            binding.button.isEnabled = true
        } else {
            binding.usernameContainer.error = reason
            binding.usernameContainer.isEnabled = true
            binding.button.text = "Sign in"
            binding.button.isEnabled = true
        }
    }

    private val mEnterPressListener = OnEditorActionListener { v: TextView, actionId: Int, _: KeyEvent? ->
        if (actionId == EditorInfo.IME_ACTION_GO) {
            onClick(v)
            return@OnEditorActionListener true
        }
        false
    }

    companion object {
        val NAME_REGEX = "[|,;]".toRegex()
        const val FRAGMENT_TAG = "sign-in-dialog"
    }
}
